In [27]:
import sympy as sym
import math
import numpy as np

**Ejercicios capitulo 3**

**2.1** Make sure SymPy is installed on your system; use it to analytically differentiate the function $f(x) = e^{sin(2x)}$. Then, evaluate the derivative at a few points from $x = 0 to x = 0.5$. Compare the latter values to the output of fprime() from code 3.1.

In [70]:
x=sym.Symbol("x", real=True)
fx = sym.exp(sym.sin(2*x))
x1=fx.diff(x)
f=sym.lambdify(x,x1)
n=np.linspace(0,5)
y=f(n*0.1)
print(y)

[2.         2.04080771 2.0815628  2.12221008 2.16269158 2.2029465
 2.2429113  2.28251962 2.32170242 2.36038793 2.39850177 2.435967
 2.47270421 2.50863159 2.54366508 2.57771848 2.61070359 2.64253037
 2.6731071  2.7023406  2.7301364  2.75639898 2.78103199 2.80393852
 2.82502133 2.84418318 2.86132708 2.87635659 2.88917619 2.89969155
 2.90780991 2.9134404  2.91649442 2.91688599 2.91453211 2.90935317
 2.90127326 2.89022059 2.87612787 2.85893263 2.83857762 2.81501116
 2.78818747 2.758067   2.7246168  2.68781075 2.6476299  2.60406268
 2.55710523 2.50676153]


In [76]:
from math import exp, sin, cos

def f(x):
       return exp(sin(2*x))
def fprime(x):
    return 2*exp(sin(2*x))*cos(2*x)
def calc_fd(f,x,h):
    fd = (f(x+h) - f(x))/h
    return fd
def calc_cd(f,x,h):
    cd = (f(x+h/2) - f(x-h/2))/h 
    return cd

if __name__== '_main_':
    x = 0.5
    an = fprime(x)

    hs = [10**(-i) for i in range(1,12)]
    fds = [abs(calc_fd(f,x,h) - an) for h in hs] 
    cds = [abs(calc_cd(f,x,h) - an) for h in hs]
    
    rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"
    print("h abs. error in fd abs. error in cd") 
    for h,fd,cd in zip(hs,fds,cds):
        print(rowf.format(h,fd,cd))


In [82]:
y=fprime(0)
y1=fprime(0.1)
y2=fprime(0.2)
y3=fprime(0.3)
y4=fprime(0.4)
y5=fprime(0.5)
print(y,y1,y2,y3,y4,y5)

2.0 2.390928390207316 2.719196694525552 2.903231667171663 2.855116148159658 2.506761534986894


**2.4** We turn to the fascinating approach of complex-step differentiation. First, Taylor
expand $f (x + ih)$ around $x$, where $f$ is an analytic function; take the imaginary part of both sides and divide by h to show that $f'(x) = Im[ f (x + ih)/h] + O(h^2)$ holds. Then, augment Fig. 3.2 to include results from this new method; you’ll need to employ complex arithmetic. Unmistakably, circumventing catastrophic cancellation pays off.

Si expandimos $f(x+ih)$ en serie de Taylor alrededor de x, obtenemos:
$f(x+ih)=f(x)+ihf'(x)+\cfrac{(ih)^2}{2}f''(x)+\cfrac{(ih)^3}{6}f'''(x)+\cdot \cdot \cdot $
$f(x+ih)=f(x)+ihf'(x)+\cfrac{-(h)^2}{2}f''(x)+\cfrac{-i(h)^3}{6}f'''(x)+\cdot \cdot \cdot $

dividiendo entre h ambos lados:

$\cfrac{f(x+ih)}{h}=\cfrac{f(x)+ihf'(x)+\cfrac{-(h)^2}{2}f''(x)+\cfrac{-i(h)^3}{6}f'''(x)+\cdot \cdot \cdot}{h} $

considerando que los terminos de orden 2 y mas grandes son desprecibles, podemos escribir:

$\cfrac{f(x+ih)}{h}=\cfrac{f(x)+ihf'(x)-O(h^2)}{h}$

Ahora si consideramos solo las partes imaginarias de las funciones.

$Im\left[\cfrac{f(x+ih)}{h}\right]=f'(x)-O(h^2) $

$f'(x)=Im\left[\cfrac{f(x+ih)}{h}\right]+O(h^2) $



